src/self_encryptor.rs

```diff
@@ -468,6 +375,110 @@ where
     Ok(())
 }
 
+async fn flush_after_write<S>(
+    state: Arc<Mutex<State<S>>>,
+    position: u64,
+    length: u64,
+) -> Result<(), SelfEncryptionError>
+where
+    S: Storage + 'static + Send + Sync,
+{
+    let old_size = {
+        let state = state.lock().await;
+        state.file_size
+    };
+
+    let new_size = cmp::max(old_size, position + length);
+
+    if new_size > old_size && old_size >= 3 * MIN_CHUNK_SIZE as u64 {
+        prepare_chunk_for_reading(Arc::clone(&state), 0).await?;
+        prepare_chunk_for_reading(Arc::clone(&state), 1).await?;
+        let mut state = state.lock().await;
+        state.chunks[0].flag_for_encryption();
+        state.chunks[1].flag_for_encryption();
+    }
+
+    if new_size < 3 * MIN_CHUNK_SIZE as u64 {
```

Comment:
this check condition won't co-exist with the above one.
your or put this block `before` the above block (as it has a internal return), or put this as an else branch.

Also, better put a commet for each code block, as this function is somehow big.